File: codigos-fonte/Python/fila.py
class No:
    def __init__(self, valor, proximo):
        self.valor = valor
        self.proximo = proximo


class Fila:
    def __init__(self):
        self.cabeca = None
        self.cauda = None

    def enfileirar(self, valor):
        novo = No(valor, None)

        if self.cabeca == self.cauda and self.cabeca == None:
            self.cabeca = self.cauda = novo
        else:
            self.cauda.proximo = novo
            self.cauda = novo
    
    def desenfileirar(self):
        if self.cabeca == self.cauda and self.cabeca == None:
            print("Não pode retirar um elemento de uma fila vazia")
            return

        temporario = self.cabeca
        if self.cabeca == self.cauda:
            self.cabeca = self.cauda = None
        else:
            self.cabeca = self.cabeca.proximo


def printAll(fila):
    atual = fila.cabeca
    while(atual != None):
        print(atual.valor, end=" ")
        atual = atual.proximo
    print()


f = Fila()
f.enfileirar(5)
printAll(f)
f.enfileirar(10)
printAll(f)
f.enfileirar(15)
printAll(f)
f.enfileirar(20)
printAll(f)
f.desenfileirar()
printAll(f)
f.desenfileirar()
printAll(f)

File: codigos-fonte/Golang/fila.go
package main

import "fmt"

type No struct {
	Valor   int
	Proximo *No
}

type Fila struct {
	Cabeca *No
	Cauda  *No
}

func (f *Fila) enfileirar(valor int) {
	novoNo := &No{
		Valor: valor,
	}
	if f.Cabeca == f.Cauda && f.Cabeca == nil {
		f.Cabeca = novoNo
		f.Cauda = novoNo
	} else {
		f.Cauda.Proximo = novoNo
		f.Cauda = novoNo
	}
}

func (f *Fila) desenfileirar() {
	if f.Cabeca == f.Cauda && f.Cabeca == nil {
		fmt.Println("Erro ao tentar retirar um elemento de uma fila vazia")
		return
	}

	if f.Cabeca == f.Cauda {
		f.Cabeca = nil
		f.Cauda = nil
	} else {
		f.Cabeca = f.Cabeca.Proximo
	}
}

func (f *Fila) printAll() {
	atual := f.Cabeca
	for atual != nil {
		fmt.Printf("%d ", atual.Valor)
		atual = atual.Proximo
	}
	fmt.Println()
}

func main() {
	f := &Fila{}
	f.enfileirar(5)
	f.printAll()
	f.enfileirar(10)
	f.printAll()
	f.enfileirar(15)
	f.printAll()
	f.enfileirar(20)
	f.printAll()
	f.desenfileirar()
	f.printAll()
	f.desenfileirar()
    f.printAll()
}
